packages/sysutils/busybox/scripts/init

```diff
@@ -570,13 +596,41 @@
         mv $UPDATE_DIR/.tmp/*/target/* $UPDATE_DIR &>/dev/null
         rm -f "$UPDATE_TAR" &>/dev/null
         rm -rf $UPDATE_DIR/.tmp &>/dev/null
-        sync
-        if [ ! -f "$UPDATE_DIR/$UPDATE_KERNEL" -o ! -f "$UPDATE_DIR/$UPDATE_SYSTEM" ] ; then
-          echo "missing ${UPDATE_KERNEL} or ${UPDATE_SYSTEM}... normal startup in 10s"
-          sync
-          usleep 10000000
-          return 0
+      elif [ -f "$UPDATE_IMG_GZ" -o -f "$UPDATE_IMG" ] ; then
+        mkdir -p $UPDATE_DIR/.tmp/mnt &>/dev/null
+
+        if [ -f "$UPDATE_IMG_GZ" ] ; then
+          echo "Found new .img.gz archive. extracting..."
+          mv "$UPDATE_IMG_GZ" $UPDATE_DIR/.tmp/update.img.gz
+          gunzip $UPDATE_DIR/.tmp/update.img.gz
+        else
+          echo "Found new .img file..."
+          mv "$UPDATE_IMG" $UPDATE_DIR/.tmp/update.img
         fi
+
+        echo "Cutting system partition from .img file..."
+        sync
+        dd if=$UPDATE_DIR/.tmp/update.img of=$UPDATE_DIR/.tmp/update.part bs=$((2048 * 512)) skip=1 conv=fsync &>/dev/null
+
+        echo "Mounting system partition..."
+        mount -o ro,loop $UPDATE_DIR/.tmp/update.part $UPDATE_DIR/.tmp/mnt
+
+        echo "Copying update files..."
+        cp $UPDATE_DIR/.tmp/mnt/KERNEL $UPDATE_DIR
```

Comment:
This needs to be:
```
mv $UPDATE_DIR/.tmp/mnt/$IMAGE_KERNEL $UPDATE_DIR/$UPDATE_KERNEL
```
as the FAT partition on RPix is:
```
-rwxr-xr-x  1 root root     12108 Mar  1 11:41 bcm2709-rpi-2-b.dtb
-rwxr-xr-x  1 root root     12750 Mar  1 11:41 bcm2710-rpi-3-b.dtb
-rwxr-xr-x  1 root root     17920 Mar  1 11:41 bootcode.bin
-rwxr-xr-x  1 root root        47 Mar  1 11:41 cmdline.txt
-rwxr-xr-x  1 root root      4785 Mar  1 11:41 config.txt
-rwxr-xr-x  1 root root      9728 Mar  1 11:41 fixup.dat
-rwxr-xr-x  1 root root   5171004 Mar  1 11:41 kernel.img
drwxr-xr-x  2 root root      8192 Mar  1 11:41 overlays
-rwxr-xr-x  1 root root   3839688 Mar  1 11:41 start.elf
-rwxr-xr-x  1 root root 112484352 Mar  1 11:41 SYSTEM
```

The different kernel naming convention may also cause a problem with the kernel md5 file - not sure what it will be called. Will it be KERNEL.md5 or kernel.img.md5?

Also, using `mv` would be a lot quicker than `cp` (same applies to the next two lines).